gluten-core/src/main/scala/org/apache/spark/util/Iterators.scala

```diff
@@ -0,0 +1,154 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.spark.util
+
+import org.apache.spark.{InterruptibleIterator, TaskContext}
+import org.apache.spark.sql.execution.metric.SQLMetric
+
+import java.util.concurrent.atomic.{AtomicBoolean, AtomicLong}
+
+private class PayloadCloser[A](in: Iterator[A])(closeCallback: A => Unit) extends Iterator[A] {
+  private var closer: Option[() => Unit] = None
+
+  TaskResources.addRecycler("Iterators#PayloadCloser", 100) {
+    tryClose()
+  }
+
+  override def hasNext: Boolean = {
+    tryClose()
+    in.hasNext
+  }
+
+  override def next(): A = {
+    val a: A = in.next()
+    closer.synchronized {
```

Comment:
I'm a bit confused why do we need `synchronized` here. Does it mean `hasNext` and `next` can be called at the same time?

Reply:
> Does it mean hasNext and next can be called at the same time?

This case is unusual... However sometime another case might be possible that the iterator was read asynchronously by another thread (say, Velox sometime hand over it's sub-tasks to background thread), but the lambda passed into `TaskResources.addRecycler` will be called from the original Spark task thread.

BTW I didn't have all the cases tested but rather programed defensively... It will be hard to debug if we actually bumped into threading issues caused by lack of locking here.